### PR TITLE
[expotools] upgrade typescript to v4.3.5

### PR DESCRIPTION
**Pull request by ajsmth**
# Why

@Simek ran into some issues with the most recent typescript update

# How

Most of the updates seemed straightforward - I've highlighted some of the reasons behind some that might not be immediately clear

# Test Plan

Rebuild with `v4.3.5` 

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: tools/src/vendoring/devmenu/steps/Task.ts [hunk 1]
```diff
@@ -61,7 +61,7 @@ export abstract class Task {
   /**
    * A function which will be call in start method. The body of the task.
    */
-  protected abstract async execute();
+  protected abstract execute();
```
Comment (ajsmth) [inline tools/src/vendoring/devmenu/steps/Task.ts:64]: Reason: 'async' modifier cannot be used with 'abstract' modifier.

File: tools/src/S3.ts [hunk 2]
```diff
@@ -198,7 +195,7 @@ export async function uploadDirectoriesAsync(
         }
       );
       let gitCommandOutput = gitCommand.stdout.toString();
-      await fsWriteFileAsync(excludeFile, gitCommandOutput);
+      fse.writeFileSync(excludeFile, gitCommandOutput);
```
Comment (ajsmth) [inline tools/src/S3.ts:198]: Reason: `promisify` seems redundant for `fse`, also is not applying right types to the wrapped function

File: tools/src/S3.ts [hunk 3]
```diff
@@ -198,7 +195,7 @@ export async function uploadDirectoriesAsync(
         }
       );
       let gitCommandOutput = gitCommand.stdout.toString();
-      await fsWriteFileAsync(excludeFile, gitCommandOutput);
+      await fse.writeFile(excludeFile, gitCommandOutput);
```
Comment (ajsmth) [inline tools/src/S3.ts:198]: Reason: promisify seems redundant for fse, also is not applying right types to the wrapped function

File: tools/src/Markdown.ts [hunk 4]
```diff
@@ -216,7 +216,7 @@ export class MarkdownRenderer implements Renderer {
       default:
         // `marked` provides much more tokens, however we don't need to go so deep.
         // So far we needed only tokens with above types.
-        throw new Error(`Cannot parse token with type: ${token!.type}`);
+        throw new Error(`Cannot parse token: ${token}`);
```
Comment (ajsmth) [inline tools/src/Markdown.ts:219]: Reason: our types don't line up perfectly with `marked`, this is now logging the full object

**Comment by Simek** [comment]
@ajsmth Thank you so much for resolving those issues! 👍 

TL/DR: Upcoming TypeDoc version uses [Recursive Conditional Types](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#recursive-conditional-types) feature which landed in TS 4.1 and bumping `tools` TS to 4.1 or newer version fixed the TypeDoc types issue, but introduces those errors in Expo codebase.

**Review by ide — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

**Pull request merged**